src/Compiler/AbstractIL/il.fs

```diff
@@ -2913,35 +2913,17 @@ and [<NoEquality; NoComparison>] ILPreTypeDef =
 
 /// This is a memory-critical class. Very many of these objects get allocated and held to represent the contents of .NET assemblies.
 and [<Sealed>] ILPreTypeDefImpl(nameSpace: string list, name: string, metadataIndex: int32, storage: ILTypeDefStored) =
-    let mutable store: ILTypeDef = Unchecked.defaultof<_>
-    let mutable storage = storage
+    let stored =
+        lazy
+        match storage with
+        | ILTypeDefStored.Given td -> td
+        | ILTypeDefStored.Computed f -> f ()
```

Comment:
I think the manual solution was in place to avoid allocations for Lazy and the closure passed into it, as per the comment above.
Also see how the "storage" was set to null to avoid holding it.

Reply:
Yes I've been thinking about it. This is almost identical to `System.Lazy<_>` implementation which also does not keep the delegate after executing it.

My reasoning was that this will not allocate any less than standard System.Lazy, so the reimplementation is not worth the cognitive overhead here.

I admit I didn't profile it. I just looked at the benchmarks in CI and concluded allocations look the same.

Reply:
We must keep the `storage` somewhere up to the point it is unwrapped, one way or another. The open question is what is the difference in performance / allocations between a member method (before) and the System.Lazy taking a delegate (this pr) and how well this gets optimized. I'll try doing a dedicated benchmark.

Reply:
The storage as a field will be still there.
But the contents of that reference, if no longer held, could have been GC'd in the original implementation and now can't - I hope I am interpreting the difference correctly.

Reply:
The point is we're not holding it at all. Once it's captured by the closure it should only be there and only for the lifetime of the delegate (?) Now I'm tempted to also decompile this and see what's what :)

Reply:
https://sharplab.io/#gist:dafe4d03aef26e424e6ac8e72350a738 This is kind of what I meant.

Anyway, this here was just a quick fix to get me unblocked.
There is value in having a carefully crafted performant implementation for memory critical stuff, I'll try to fix the current implementation and compare properly.

Reply:
The arguments are passed as part of the closure to Lazy. But if Lazy gets rid of the closure object, it will work the same.
https://sharplab.io/#v2:DYLgZgzgNAJiDUAfAsAKDQFwJ4AcCmABAJIAyAKrngCJ5gEC8BEGATgJYB2A5mgNoA8AOQD2AUQCOAVwCGwNtgDcIgMLCAtjmnsIwjgD4AupkrFylGmADKGYSzwwGBVRskZ7BYXQAUkjvIIAtHqmFPgWAJRoxvgEApZ4svaGpgAKdqHUtEQawF4c0mp4lpoAxnggTKycXARyzFAE+YUVzOzcDYUY0jDSXUQcMHgAHhWcGADMAEwNzLbSXOUh5rTWtvbhDGgE27V4GJVrDvRbO6fA0gBeWCen22q9JQAWByzzhADu8o83t4hLYSsbHYYAA6Zw4VzuOhBAjeSKoW47H6FNQAIzwLAIAH0QQBxPZeDaMWbAkEANVkkjwUXQtKAA
OK, I am convinced - will approve.